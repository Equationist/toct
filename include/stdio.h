/* stdio.h - Standard Input/Output Library for CompilerKit C Compiler */

#ifndef _STDIO_H
#define _STDIO_H

/* Standard I/O types */
typedef struct {
    int _fd;
    int _flags;
    char* _buffer;
} FILE;

/* Standard streams */
extern FILE* stdin;
extern FILE* stdout; 
extern FILE* stderr;

/* Standard I/O functions */
int printf(const char* format, ...);
int fprintf(FILE* stream, const char* format, ...);
int sprintf(char* str, const char* format, ...);
int snprintf(char* str, int size, const char* format, ...);

int scanf(const char* format, ...);
int fscanf(FILE* stream, const char* format, ...);
int sscanf(const char* str, const char* format, ...);

int putchar(int c);
int puts(const char* s);
int fputs(const char* s, FILE* stream);
int fputc(int c, FILE* stream);

int getchar(void);
char* gets(char* s);
char* fgets(char* s, int size, FILE* stream);
int fgetc(FILE* stream);
int ungetc(int c, FILE* stream);

FILE* fopen(const char* filename, const char* mode);
FILE* freopen(const char* filename, const char* mode, FILE* stream);
int fclose(FILE* stream);
int fflush(FILE* stream);

long ftell(FILE* stream);
int fseek(FILE* stream, long offset, int whence);
void rewind(FILE* stream);
int feof(FILE* stream);
int ferror(FILE* stream);
void clearerr(FILE* stream);

/* File position constants */
#define SEEK_SET 0
#define SEEK_CUR 1
#define SEEK_END 2

/* End of file constant */
#define EOF (-1)

/* Buffer size */
#define BUFSIZ 8192

/* Null pointer */
#ifndef NULL
#define NULL ((void*)0)
#endif

#endif /* _STDIO_H */