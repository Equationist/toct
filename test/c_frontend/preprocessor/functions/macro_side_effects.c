/* Test macro side effects and multiple evaluation */
#define UNSAFE_MAX(a, b) ((a) > (b) ? (a) : (b))
#define UNSAFE_SQUARE(x) ((x) * (x))

#include <stdio.h>

int increment_counter() {
    static int counter = 0;
    return ++counter;
}

int main() {
    int x = 5, y = 3;
    
    printf("Initial x = %d, y = %d\n", x, y);
    
    /* This will increment x twice due to macro expansion */
    printf("UNSAFE_MAX(++x, y) = %d\n", UNSAFE_MAX(++x, y));
    printf("x after UNSAFE_MAX = %d\n", x);
    
    /* Reset and test with function call */
    x = 5;
    printf("Reset x = %d\n", x);
    printf("UNSAFE_SQUARE(increment_counter()) = %d\n", UNSAFE_SQUARE(increment_counter()));
    printf("Counter called multiple times due to macro expansion\n");
    
    return 0;
}