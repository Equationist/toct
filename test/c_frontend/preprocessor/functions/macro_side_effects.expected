Initial x = 5, y = 3
UNSAFE_MAX(++x, y) = 7
x after UNSAFE_MAX = 7
Reset x = 5
UNSAFE_SQUARE(increment_counter()) = 2
Counter called multiple times due to macro expansion